(GH-2183) Document new module workflow (#2204)
puppetlabs/bolt: documentation/managing_modules.md

> # 🧪 Managing modules in Bolt projects

Reviewer: I also wonder if adding a "Why we did this" or "Benefits of using this workflow" type of section would be useful. No more accidentally wiping modules in`modules/`, no more having to manually list and resolve dependencies in the Puppetfile.

Author: I can put that in for now, but I think that's the kind of info that should go in a developer update (perhaps along with the "What's changed" stuff). Reason being, I'll probably remove a lot of that sort of info once this properly replaces the old way of doing things.

Author: Also, this is kind of what happened with Projects - I like the idea of a more permanent home for "Why we're doing stuff"

Bolt version 2.30.0 introduced module workflows for Bolt projects that improve
dependency management and make it easier to add new modules to your Bolt
project. These workflows, which include the `bolt module install`, `bolt module
add`, and `bolt project migrate` commands, are experimental.

> ## What's changed?

Reviewer 2: I think this should go after "How it works" so the user has a bit more context for these changes.

Reviewer: Hm, I kind of disagree. I think most users reading this will already use Bolt and will be most interested this section, and it helps define the scope of what's changed so that reading on feels less intimidating.

Author: I actually originally had this section after How it works, but I didn't like how it separated "How it works" from "opting in" etc.

- **Module installation directory:** Bolt now installs modules into the
  `.modules/` directory instead of `modules/`. Avoid committing `.modules/` to
  source control. Your users can download your Bolt project and use the `bolt
  module install` command to download the required modules from the forge.
- **Module configuration:** Your `bolt-project.yaml` file contains a `modules`
  key that lists the direct module dependencies of your Bolt project. Unless you
  need to pin a module to a specific version, avoid editing the `modules` list
  directly. Instead, use the `bolt module add` command.

> - **Puppetfile:** The Puppet file is now a lock file. Bolt generates a

Reviewer: ```suggestion
- **Puppetfile:** The Puppetfile is now a [lock file](https://bundler.io/rationale.html#checking-your-code-into-version-control). Bolt generates a
```

Reviewer: Or another link that explains what a lockfile is

Author: I'm going to add links to `Gemfile.lock` and `yarn.lock` below this

  Puppetfile each time you modify your modules with a Bolt command. Do not edit
  the Puppetfile directly. Instead, use Bolt commands to manage your modules,
  and rely on Bolt to manage the Puppetfile. You can compare a Puppetfile to a
  `Gemfile.lock` file in Ruby, or a `yarn.lock` file in Yarn.

> - **Module path:** The new module path is `['modules']` and Bolt always appends

Reviewer: Looks like Puppet uses `modulepath` https://puppet.com/docs/puppet/6.18/dirs_modulepath.html we seem to use `module path` and `modulepath` pretty much interchangeably in the Bolt docs. I'm fine with either, it'd be good to standardize.

Author: Yeah, good point. There are a couple of other things we should standardize too. I'll add this to the list and open a ticket for those.

Reviewer: https://github.com/puppetlabs/bolt/pull/2225 Should do it.

  `.modules` to the module path.

## How it works

When you install a Forge module, Bolt 
- installs the latest version of the module along with any dependencies required
  by the module.
- adds entries for the module to the `modules` key in your `bolt-project.yaml`,
  and installs the module and its dependencies into a `.modules` directory in
  your project. 
- generates a Puppetfile, which lists all of a project's required modules and

>   their versions.

Author: Ah yes, thank you. I was trying to be more general for sure. I don't want to introduce specific commands in this section.

> For example, if you create a project named `example_project` and install the

Reviewer: ```suggestion
For example, if you create a project named `example_project` with `bolt project init` and install the
```

> `apache` module, the `modules` key in your `bolt-project.yaml` file looks

Reviewer: ```suggestion
`apache` module with `bolt module add puppetlabs-apache`, the `modules` key in your `bolt-project.yaml` file looks
```

something like this:

```yaml
name: example_project
modules:
- name: puppetlabs-apache
```

The project's Puppetfile lists the `apache` module and all of its
dependencies:

```puppet
# This Puppetfile is managed by Bolt. Do not edit.
mod "puppetlabs-apache", "5.5.0"
mod "puppetlabs-stdlib", "6.4.0"
mod "puppetlabs-concat", "6.2.0"
mod "puppetlabs-translate", "2.2.0"
```

And the `.modules` directory contains all of the installed modules:

```shell

> $ ls modules

Reviewer 2: ```suggestion
$ ls .modules
```

apache    concat    stdlib    translate
```

## Opting in

> Use `bolt project migrate` to migrate a Bolt project to the new workflows. 

Reviewer 2: `bolt project migrate` won't always work, or a user may not want to use the command for their own reasons. We should have a section describing how to manually opt in. We should recommend the user be using a VCS for their project. Assuming you have a Bolt project with a `bolt-project.yaml`, these are the steps they'll need to take:

- Add a `modules` key to the `bolt-project.yaml`
- Add any modules you want to manage with Bolt from your Puppetfile to the `modules` key. Only list modules that you use directly and not their dependencies. Only Puppet Forge modules are supported.
- Delete the Puppetfile.
- Delete the modules you are managing, plus their dependencies, from the modulepath (this is likely `modules` for most users).
- Move all remaining modules (likely in `site-modules` / `site`) to the `modules` directory.
- If you have configured `modulepath` in `bolt-project.yaml`, unconfigure it.
- Run `bolt module install` to install the managed modules.

Author: I was trying to avoid describing this workflow, because it seems like a less happy path than just using the `migrate` and `add` commands, and it makes it more difficult to be really clear about what a user needs to do to get going. 

I think we often fall into the trap of trying to document every possible way of doing something. Could you give me an example of a case where the migrate command won't work? Obviously, we want to document it if it's going to be an issue for a large group of users, but maybe it's a case of covering that use case in a separate doc?

Reviewer 2: I think the most likely scenario is if an existing Puppetfile cannot be parsed, preventing a user from using `project migrate`. Per the chat in channel, I think we can hold off on describing this until/if users keep asking for it.

Bolt reads your Puppetfile and prompts you to choose which modules are direct
dependencies of your project. Only select a module as a direct dependency if
your project uses content from that module.

After you've selected your project's direct dependencies, Bolt adds those
modules to your `bolt-project.yaml` file and pins them to their current
versions, installs those modules and their dependencies into a `.modules`
directory, and generates a Puppetfile. 

If you don't want your modules pinned to a specific version, edit out the
`version_requirement` fields for each module in your `bolt-project.yaml` file
and run `bolt module install`.

> **NOTE:** Migrating a project makes irreversible changes to the project's
  configuration and inventory files. Before you migrate, make sure the project
  has a backup or uses a version control system.

## Install modules in a Bolt project

> Once you've opted in to the experimental workflows, you can use the `bolt module

Reviewer 2: ```suggestion
Once you've opted in, you can use the `bolt module
```

> add` and `bolt module install` commands to manage your Forge modules. Git
> modules are not currently supported.

Reviewer 2: ```suggestion
add` and `bolt module install` commands to manage your Forge modules.

> **Note:** Bolt only supports managing modules from the public Puppet
> Forge. Git modules and private Forge modules are not supported.
```

This is a fairly important point that might trip up more advanced users. Having it in a blockquote should hopefully make it a bit more noticeable.

### Add a Forge module and its dependencies to an existing project

To add a single module and its dependencies to your Bolt project, use the 
`bolt module add` command. For example, to add the puppetlabs/apt module:

*\*nix shell command*
```shell
bolt module add puppetlabs/apt
```

*Powershell cmdlet*
```shell
Add-BoltModule -Module puppetlabs/apt
```

Bolt adds a declaration to your project configuration file, resolves
dependencies, updates your project's Puppetfile, and installs the modules to the
project's `.modules` directory.

> ### Install the modules associated with a Bolt project

Reviewer: Might be worth noting this is also the command you would use to update module versions to latest.

If you've just downloaded a Bolt project from source control and want to install
the modules associated with it, use the following command:

*\*nix shell command*
```shell
bolt module install
```

*Powershell cmdlet*
```shell
Install-BoltModule
```

Bolt loads your `bolt-project.yaml` file, resolves dependencies for all of the
declared modules, writes a Puppetfile to your project that includes
specifications for the resolved modules, and installs the modules to the
project's `.modules` directory.

> ## How Bolt resolves shared dependencies

Reviewer: ```suggestion
## How Bolt resolves dependencies
```

> ## How Bolt resolves shared dependencies

Reviewer 2: This bit only applies to `bolt module add`. I think it gives a little more information than the user needs, so maybe something simpler under "Add a Forge module and its dependencies..." like this?

```
When Bolt adds a new module to the project and resolves its dependencies, it 
attempts to keep all installed modules on the same versions. If Bolt is unable to
do this due to a version conflict, it may update installed modules to newer versions.
```

Each time you add a new module to a project, Bolt attempts to resolve all of the
modules in the project configuration file to ensure that there are no version
conflicts between modules and their dependencies.

If a new module shares a dependency with another module, Bolt might need to
install a different version of the dependency that is compatible with both
modules. When adding a new module, Bolt attempts to change the version of as few
modules as possible. Bolt resolves modules in the following order:
1. Pin all installed modules to a specific version and resolve with the new
   module. Does not modify the version of any installed module.
2. Unpin all installed modules and resolve with the new module. Update all
   modules to the latest compatible version.

> If Bolt successfully resolves modules at any step, it does not continue to the

Reviewer: ```suggestion
If Bolt fails to resolve modules at any step, it does not continue to the
```

Reviewer: I think?

next step.

> ## Forcibly installing modules

Reviewer 2: Likewise, this only applies to `bolt module install`, so maybe having it rolled into that section?

  and rely on Bolt to manage the Puppetfile. You can compare a Puppetfile to a
  [`Gemfile.lock`](https://bundler.io/rationale.html#checking-your-code-into-version-control)
  file in Ruby, or a
  [`yarn.lock`](https://classic.yarnpkg.com/en/docs/yarn-lock/) file in Yarn.
- **modulepath:** The new modulepath is `['modules']` and Bolt always appends
  `.modules` to the modulepath.

## How it works

A Bolt project lists its module dependencies in the `bolt-project.yaml` file
under the `modules` key. For example, a Bolt project that uses the `mysql` and
`apache` modules might look like this:

```yaml
modules:
- puppetlabs/apache
- puppetlabs/mysql
```

When you use Bolt to install a project's modules, it loads the information from
the `modules` key in your `bolt-project.yaml`, resolves the modules and their
dependencies, generates a Puppetfile listing all of the modules to install, and
then installs the modules.

If your project needs another module, you can use Bolt to add the module to your
project configuration, generate a new Puppetfile that includes the new module
and its dependencies, and install the modules.

## Opting in

> **Before you begin:** To opt in, you need a Bolt project. You can crate a new

Reviewer 2: ```suggestion
**Before you begin:** To opt in, you need a Bolt project. You can create a new
```

project with some Forge modules using the steps described in [Installing
modules](bolt_installing_modules.md#create-a-new-bolt-project-and-install-a-list-of-modules-with-dependencies),
or use an existing Bolt project. If you choose to use an existing Bolt project,
**create a backup first**!

Use `bolt project migrate` to migrate a Bolt project.

Bolt reads your Puppetfile and prompts you to choose which modules are direct
dependencies of your project. Only select a module as a direct dependency if
your project uses content from that module.

After you've selected your project's direct dependencies, Bolt adds those
modules to your `bolt-project.yaml` file and pins them to their current
versions, installs those modules and their dependencies into a `.modules`
directory, and generates a Puppetfile. 

If you don't want your modules pinned to a specific version, edit out the
`version_requirement` fields for each module in your `bolt-project.yaml` file

> and run `bolt module install`.

Reviewer 2: ```suggestion
and run `bolt module install --force`.
```

If the Puppetfile is missing any module declarations, Bolt assumes the
Puppetfile is not managed by Bolt and raises an error like this:

```shell
Puppetfile /myproject/Puppetfile is missing specifications for the following
module declarations:

- name: puppetlabs-ruby-task-helper

This may not be a Puppetfile managed by Bolt. To forcibly overwrite the
Puppetfile, run 'bolt module install --force'.
```

This error usually only occurs if you've manually modified your
`bolt-project.yaml` or `Puppetfile`. To resolve the conflict and install the
modules declared in `bolt-project.yaml`, run the following command:

*\*nix shell command*
```shell
bolt module install --force 
```

*Powershell cmdlet*
```shell
Install-BoltModule -Force
```

## Pin a module version

If you need to pin a module in your Bolt project to a specific version, you can
add a version requirement to the module in your `bolt-project.yaml` and run
`bolt module install --force`.

Follow these steps to pin a module version:
1. Find the module under the `modules` key in your `bolt-project.yaml` file.
2. Add a `version_requirement` key that specifies the version requirements for

>    the module. For example, the following example from a `bolt-project.yaml`

Reviewer 2: ```suggestion
   the module. For example, the following `bolt-project.yaml`
```

>    file sets the Ruby Task Helper module to greater than or equal to 1.0.0, but
>    less than 3.0.0:

Reviewer 2: ```suggestion
   file sets the version requirement for the `apache` module to greater than or equal
    to 4.0.0, but less than 6.0.0:
```

     version_requirement: '>= 4.0.0 < 6.0.0'
   ```
3. Run `bolt module install --force`. The `--force` flag is required because
   you've made a change to your `bolt-project.yaml` file, and it no longer
   matches the Puppetfile.

When you run `bolt module install --force`, Bolt attempts to resolve all of the
modules in the project configuration file to ensure that there are no version
conflicts between modules and their dependencies. Bolt fails with an error
message if it cannot resolve a dependency due to a version requirement. 

> 🔩 **Tip**: For information on how to specify module versions, see the Puppet
> documentation on [Specifying
> versions](https://puppet.com/docs/puppet/latest/modules_metadata.html#specifying-versions).

## Update your project's modules

To update the modules in your project, run:

*\*nix shell command*
```shell
bolt module install --force 
```

*Powershell cmdlet*
```shell
Install-BoltModule -Force
```

Bolt updates your modules to the latest available versions and attempts to
resolve all of the modules in the project configuration file to ensure that
there are no version conflicts between modules and their dependencies.

## Examples

> ### Example: Migrating an existing Bolt project

Reviewer 2: ```suggestion
### Migrating an existing Bolt project
```

The extra "Example" seems redundant.

Author: Yeah, these are kinda gross. I'm a little afraid people will navigate straight to these and get confused. But maybe I need to have more faith 😄

mod "puppetlabs-puppetserver_gem", "1.1.1"
```

If you ran the `migrate` command, and selected the `apache`, `apt`, and `mysql` modules as direct dependencies of your Bolt project, Bolt would do the following:
- Update your `bolt-project.yaml` file to add the `modules` key, together with the `apache`, `apt`, and `mysql` modules:
  ```yaml
  ---
  name: myproject
  modules:
  - name: puppetlabs-apache
  version_requirement: "=5.5.0"
  - name: puppetlabs-apt
  version_requirement: "=7.6.0"
  - name: puppetlabs-mysql
  version_requirement: "=10.7.1"
  ```
- Resolve your dependencies and generate a new Puppetfile:
  ```puppet
  # This Puppetfile is managed by Bolt. Do not edit.
  mod "puppetlabs-apache", "5.5.0"
  mod "puppetlabs-apt", "7.6.0"
  mod "puppetlabs-mysql", "10.7.1"
  mod "puppetlabs-stdlib", "6.4.0"
  mod "puppetlabs-concat", "6.2.0"
  mod "puppetlabs-translate", "2.2.0"
  mod "puppetlabs-resource_api", "1.1.0"
  mod "puppetlabs-puppetserver_gem", "1.1.1"
  ```
- Install the modules from the Puppetfile into the `.modules` directory.
- Remove the old managed modules from the `modules/` directory.
- Move `mymodule` from `site-modules/` to `modules/`.

> ### Example: Install a module with `add`

Reviewer 2: ```suggestion
### Install a module with `add`
```

module add` command. For example, to add the puppetlabs/apt module:

*\*nix shell command*
```shell
bolt module add puppetlabs/apt
```

*Powershell cmdlet*
```shell
Add-BoltModule -Module puppetlabs/apt
```

Bolt adds a declaration to your project configuration file, resolves
dependencies, updates your project's Puppetfile, and installs the modules to the
project's `.modules` directory.

When Bolt adds a new module to the project and resolves its dependencies, it
attempts to keep all installed modules on the same versions. If Bolt is unable
to do this due to a version conflict, it may update installed modules to newer
versions.

### Install the modules associated with a Bolt project

If you've just downloaded a Bolt project from source control and want to install
the modules associated with it, use the following command:

*\*nix shell command*
```shell
bolt module install
```

*Powershell cmdlet*
```shell
Install-BoltModule
```

Bolt loads your `bolt-project.yaml` file, resolves dependencies for all of the
declared modules, writes a Puppetfile to your project that includes
specifications for the resolved modules, and installs the modules to the
project's `.modules` directory.

> #### Forcibly installing modules

Author: ```suggestion
#### Forcibly install modules
```

   the module. For example, the following `bolt-project.yaml` file sets the
   version requirement for the `apache` module to greater than or equal to
   4.0.0, but less than 6.0.0:
   ```yaml
   # bolt-project.yaml
   ...
   modules:
   - name: puppetlabs/apache
     version_requirement: '>= 4.0.0 < 6.0.0'
   ```
3. Run the following command. The `force` option is required because
   you've made a change to your `bolt-project.yaml` file, and it no longer
   matches the Puppetfile.
   
   *\*nix shell command*
   ```shell
   bolt module install --force 
   ```

   *Powershell cmdlet*
   ```shell
   Install-BoltModule -Force
   ```

When you run the install command with the `force` option, Bolt attempts to resolve all of the
modules in the project configuration file to ensure that there are no version
conflicts between modules and their dependencies. Bolt fails with an error
message if it cannot resolve a dependency due to a version requirement. 

> 🔩 **Tip**: For information on how to specify module versions, see the Puppet
> documentation on [Specifying
> versions](https://puppet.com/docs/puppet/latest/modules_metadata.html#specifying-versions).

## Examples

> ### Migrating an existing Bolt project

Author: ```suggestion
### Migrate an existing Bolt project
```
